WeatherLocator ipAddress caching in config.yml

adrienzag opened the pull request:
- adjusted ServerLocator class to allow for instance of WeatherByLocation class to be injected through constructor, injected instance in WeatherByLocation.java (line 103).

- adjusted ServerLocator class to cache server Location data to config.yml anytime that ipAddress changes detected.

adrienzag committed:
adjusted ServerLocator class to allow for instance of WeatherByLocation class to be injected through constructor, injected instance in WeatherByLocation.java.

adjusted ServerLocator class to cache server Location data to config.yml anytime that ipAddress changes detected.

scottbarnesg commented:
Thanks for submitting this PR! I gave this an initial one-over and it looks pretty good. Will pull and test soon.

scottbarnesg commented on .idea/misc.xml:
This file should not be pushed. My bad for not adding it to the `.gitignore` file.

Could you revert the changes to this file and add it to the `.gitignore`?

adrienzag replied:
understood, will do

scottbarnesg commented on src/main/java/mc/spigot/weatherbylocation/ServerLocator.java:
```diff
@@ -70,6 +84,32 @@ private static LocationData geolocateIpAddress(String ipAddress) throws IOExcept
         locationResult.longitude = jsonNode.get("lon").asDouble();
+
+        // Save new locationData object to config.yml file
+        config.set("ipAddress", ipAddress);
```
Let's break out these `config.set()` calls into their own function. Maybe called something like `saveDataToConfig()`?

Then call the function here and pass in `ipAddress` and `locationResult`.

adrienzag replied:
Understood, the block that sets values to the config file are to be replaced by a function call.

Since this is inside the geolocateIpAddress() function that passes "FileConfiguration config" as one of its parameters is it safe to assume that either the saveDataToConfig() method should be overloaded to take a 3rd parameter (config) in order to avoid making multiple calls to this.plugin.getConfig(); or that it should just always have at least the 3 parameters ? (ipAddress, locationResult, config)

scottbarnesg replied:
I think it just makes sense to call `plugin.getConfig()` in the function. I don't think there's really much cost to calling it multiple times.

adrienzag replied:
Ok duly noted, will revert the recent commit and adjust

thanks

adrienzag replied:
Upon giving it another look, I am not sure if geolocateIpAddress() should still be a static method ? 

If it is to make a call to the function saveDataToConfig() we are presented with the following issues :  

- if saveDataToConfig() is non-static then it will need to make a call to get instance of config with this.plugins.getConfig() , but geolocateIpAddress() cannot make a call to the instance-specific function as it is static 

- if saveDataToConfig() is static then we would not be using class instances, and so the most straightforward way to get access to the config.yml file would be to pass it as a parameter - which seems to not be something desired as of now

I guess it also would depend on how geolocateIpAddress() would be used but I am not sure I am able to see why static is better than non static here.

 Alternatively, maybe saveDataToConfig() is better off in the locate() function within the class

adrienzag committed:
Revert "adjusted ServerLocator class to allow for instance of WeatherByLocation class to be injected through constructor, injected instance in WeatherByLocation.java."

This reverts commit a8773397c8bd1de65ca74b4fc11ab9c9a0fe225d.

adrienzag committed:
adjusted ServerLocator class to allow for instance of WeatherByLocation class to be injected through constructor, injected instance in WeatherByLocation.java.

adjusted ServerLocator class to cache server Location data to config.yml anytime that ipAddress changes detected.

adrienzag committed:
Revert "adjusted ServerLocator class to allow for instance of WeatherByLocation class to be injected through constructor, injected instance in WeatherByLocation.java."

This reverts commit 4a1f0e40272ad284b412881a11f78074551b0ca3.

scottbarnesg commented on src/main/java/mc/spigot/weatherbylocation/ServerLocator.java:
```diff
@@ -70,8 +84,43 @@ private static LocationData geolocateIpAddress(String ipAddress) throws IOExcept
+        config.set("city",  locationResult.city);
+        config.set("latitude", locationResult.latitude);
+        config.set("longitude",  locationResult.longitude);
+
```
I needed to add this to get the plugin to save these values to the config file.

```suggestion
plugin.saveConfig();
```

adrienzag replied:
Aha i see, sorry about that mistake. I knew that Bukkit had an in-memory copy of the config file, but it seems that in order for it to be written to disk (the config.yml) we must explicitly save it as it does not automagically happen  [https://bukkit.fandom.com/wiki/Configuration_API_Reference]

In the "tests" i was running, i was fetching the values with getXXX methods and they were showing up (but obviously this came from memory) i was wondering why the config.yml file seemed to be "erased" after i closed the server.

Will not happen again

scottbarnesg replied:
No worries! I made the same mistake when first implementing the configuration file.

scottbarnesg commented on src/main/java/mc/spigot/weatherbylocation/ServerLocator.java:
```diff
@@ -70,8 +84,43 @@ private static LocationData geolocateIpAddress(String ipAddress) throws IOExcept
         return locationResult;
     }
+
+    private static void saveDataToConfig(String ipAddress, LocationData locationResult, FileConfiguration config) {
```
Can't be static not that it calls `plugin.saveConfig();`
```suggestion
    private void saveDataToConfig(String ipAddress, LocationData locationResult, FileConfiguration config) {
```

scottbarnesg commented on src/main/java/mc/spigot/weatherbylocation/ServerLocator.java:
```diff
@@ -45,13 +59,13 @@ private static String fetchIpAddress() throws IOException, InterruptedException
     }
 
-    private static LocationData geolocateIpAddress(String ipAddress) throws IOException, InterruptedException {
+    private static LocationData geolocateIpAddress(String ipAddress, FileConfiguration config) throws IOException, InterruptedException {
```
Can't be static, per other changes.
```suggestion
    private LocationData geolocateIpAddress(String ipAddress, FileConfiguration config) throws IOException, InterruptedException {
```

scottbarnesg commented:
@adrien2-public I ran your code on my test server, and made a few comments based on the results.

Specifically, I had to add `plugin.saveConfig()` to  `saveDataToConfig()` so that it would write the updates to the config file.

adrienzag committed:
adjusted code to allow for saving of location data to disk on config.yml when IP address is changed

adrienzag commented:
Made the changes to the files, it seems that i had been testing wrong as in the above reply mentioned.

Most recent changes include the below:
 - the Bukit Configuration API Reference says that it is generally preferable to use getConfig() instead of assigning it to an instance variable. So in the spirit of this i replaced most places that refered to "config" as a variable to plugin.getConfig() (excluding the locate() function) and i removed use of config as a method parameter as they are no longer static methods.

- made locationDataFromConfig() into non-static method so we can refer to the config.yml file without passing a parameter value but rather use "plugin.getConfig()"

- made some comments in the saveDataToConfig() method that distinguishes the setting of values to config.yml and the saving of values to disk on config.yml. Wanted to write a log comment at the end of function that would say "Location data for IP XXX.XXX.XXX saved to disk on xx:xx:xx  xx/xx/xxxx but left it at that since I am not sure if this is desired.

scottbarnesg reviewed (approved):
Pulled and confirmed this works on my test server. Looks good to me!
